---
title: "FigureYa158MutationPattern"
author: "小丫画图出品"
date: "2019-12-29"
output: html_document
---
欢迎关注“小丫画图”公众号，回复“小白”，看小视频，实现点鼠标跑代码。

小丫微信: epigenomics  E-mail: figureya@126.com

作者：SRF

小丫编辑校验

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述

输入TCGA突变数据，用非负矩阵分解方法NMF分析突变特征，重复出如下图。

![](example.png)

出自<https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4610178/>

# 应用场景

突变特征定义：体细胞突变是多个突变过程如DNA修复缺陷，暴露于外源或内源诱变剂等综合结果。不同的突变过程会产生不同组合的突变类型，即突变特征。利用突变特征分析感兴趣癌种突变特征，探讨其可能的发生病因及机制。
目前已经有很多的R包可以帮助我们完成mutation signature分析如MutationalPatterns,maftools,deconstructSigs ,SomaticSignatures.

在这里我们选用maftools + MutationalPatterns来完成分析。

# 环境设置

使用国内镜像安装包

```{r, eval=FALSE}
options("repos"= c(CRAN="https://mirrors.tuna.tsinghua.edu.cn/CRAN/"))
options(BioC_mirror="http://mirrors.ustc.edu.cn/bioc/")
BiocManager::install("MutationalPatterns")
BiocManager::install("BSgenome")
BiocManager::install("BSgenome.Hsapiens.UCSC.hg38")
install.packages("tidyverse")
install.packages("haven")
```

加载包

```{r}
library(MutationalPatterns)
library(TCGAbiolinks)
library(maftools)
library(magrittr)
library(stringr)
library(NMF)
library(ggsci)
library(BSgenome)
library(ggplot2)
library(cowplot)
library(BSgenome.Hsapiens.UCSC.hg38)
Sys.setenv(LANGUAGE = "en") #显示英文报错信息
options(stringsAsFactors = FALSE) #禁止chr转成factor
```

# 输入文件的下载

用TCGAbiolinks下载突变信息、临床数据

```{r}
## 突变数据
maf_BLCA <- GDCquery_Maf("BLCA", save.csv = FALSE, directory = "GDCdata", pipelines = "mutect2") 
#下载一次就好，保存到文件，便于后面加载
save(maf_BLCA, "maf_BLCA.RData")

## clinical data 
blca_clinical <- TCGAquery_subtype(tumor = "blca")
#非肌层浸润性膀胱癌是指局限于膀胱黏膜层（Tis、Ta）及固有层（T1）
table(blca_clinical$`AJCC Tumor category`)
mibc_samples <- blca_clinical[blca_clinical$`AJCC Tumor category` %in%
                                c("T2", "T3", "T3b", "T4", "T4a"), ]$patient
```

# NMF分析

```{r}
# 载入前面下载的MIBC_mutation
load("maf_BLCA.RData")
maf_BLCA <- read.maf(maf_BLCA, isTCGA = T) %>% subsetMaf(., tsb = mibc_samples, isTCGA = TRUE)
#96 mutational profile：4*6*4=96
#这里useSyn选的不包含同义突变，要包含参数选T
#包含useSyn画出来的mutation signature跟原文更相似。
mut_mat <- trinucleotideMatrix(maf = maf_BLCA, 
                               ref_genome = "BSgenome.Hsapiens.UCSC.hg38",
                               prefix = NULL, ignoreChr = NULL, 
                               useSyn = FALSE) %>% .$nmf_matrix %>% t()

#De novo mutational signature extraction using NMF
#用于自己的数据run最好在30-50之间，为了加快速度这里用10
mut_mat <- mut_mat + 0.0001
estimate <- nmf(mut_mat, rank=2:10, method="brunet", nrun=10, seed=123456)#运行时间久
plot(estimate)

#最佳个数的选择，一般选择cophenetic correlation开始下降的点，轮廓系数做大的点，详情请仔细阅读NMF说明文档。
#选择5作为最佳点,extract_signatures设置了seed:123456
nmf_res <- extract_signatures(mut_mat, rank = 5, nrun = 30)
colnames(nmf_res$signatures) <- paste0("Signature", LETTERS[1:5])
rownames(nmf_res$contribution) <- paste0("Signature", LETTERS[1:5])
```

# 跟COSMIC突变特征比较

[COSMIC mutation signature](https://cancer.sanger.ac.uk/cancergenome/assets/signatures_probabilities.txt)

**V2版本**可直接下载到COSMIC归纳的30种signature.

**V3版本**在2019年5月份分布,对突变的类型进行了进一步的细分，分为Single Base Substitution (SBS) Signatures,Doublet Base Substitution (DBS) Signatures,Small Insertion and Deletion (ID) Signatures。
这里我们用V2版本的来对signature进行注释。

```{r}
#cancer signature
sp_url <- paste("https://cancer.sanger.ac.uk/cancergenome/assets/",
           "signatures_probabilities.txt", sep = "")
cancer_signatures = read.table(sp_url, sep = "\t", header = TRUE)
new_order = match(row.names(mut_mat), cancer_signatures$Somatic.Mutation.Type)
cancer_signatures = cancer_signatures[as.vector(new_order),]
row.names(cancer_signatures) = cancer_signatures$Somatic.Mutation.Type
cancer_signatures = as.matrix(cancer_signatures[,4:33])
all(rownames(cancer_signatures) == rownames(nmf_res$signatures))
cos_sim_signatures = cos_sim_matrix(nmf_res$signatures, cancer_signatures)

#跟已经发表的cosmic数据库signature比较，注释mutation signature
plot_cosine_heatmap(cos_sim_signatures,
                    cluster_rows = FALSE, plot_values = T)
#重命名signatures
colnames(nmf_res$signatures) <- c("C>T_CpG", "APOBEC1", "Unknow",  "POLE", "APOBEC2")
rownames(nmf_res$contribution) <- c("C>T_CpG", "APOBEC1", "Unknow",  "POLE", "APOBEC2")

# 保存到文件，便于后面修改画图代码时直接载入
save(nmf_res, file = "nmf_res.RData") 
```

[COSMIC](https://cancer.sanger.ac.uk/cosmic/signatures_v2)提供了每个signature的详细信息，我们可以发现SignatureA-E分别与COSMIC数据库的Signature1,13,5,10,2最相似即分别是:age,APOBEC,Unknow,POLE,APOBEC
文中提到的signature1是C>T_CpG，查看文章[Signatures of mutational processes in human cancer](https://www.nature.com/articles/nature12477)也有提到Signature 1A/B is probably
related to the relatively elevated rate of spontaneous deamination of 5-methyl-cytosine which results in C.T transitions and which predominantly occurs at NpCpG trinucleotides9.因此命名方面可以多加参考。

# 开始画图

**MutationalPatterns**这个包有两个函数plot_96_profile和plot_contribution，可以直接出图，但是可选的参数很少并且看代码也就是重新包装了一下ggplot2。

下面的绘图，我们就借鉴其数据处理方法，然后用ggplot2画出图，这样可以个性化的调整细节。

```{r}
col <- pal_npg("nrc", alpha=0.7)(6) 
names(col) <- c("C->T", "T->C", "C->A", "C->G", "T->A", "T->G")

## Figure1A
load("nmf_res.RData")
signature.col <- c("red", "yellow", "darkorchid2", "steelblue2", "black")
names(signature.col) <- c("APOBEC1", "APOBEC2", "C>T_CpG", "POLE", "Unknow")
#mutation pattern输入数据准备
Pattern_input <- function(mut_matrix){
  CONTEXTS_96 <- rownames(mut_matrix)
  substitution <- str_sub(rownames(mut_matrix), 3, 5)
  substitution <- str_replace(substitution, ">", "->")
  context = CONTEXTS_96
  context = str_replace(context, "\\[.*]", "_")
  df = data.frame(substitution = substitution, context = context)
  df2 = cbind(df, as.data.frame(mut_matrix))
  df3 = reshape2::melt(df2, id.vars = c("substitution", "context"))
  return(df3)
}
pattern.dat <- Pattern_input(nmf_res$signatures)
pattern.dat$variable <- factor(pattern.dat$variable,
                               levels = c("APOBEC1", "APOBEC2", "C>T_CpG", "POLE", "Unknow"))

p1 <- ggplot(data = pattern.dat, aes(x = context, y = value, 
                       fill = substitution, width = 1)) + 
  geom_bar(stat = "identity", colour = "black", size = 0.2) + 
  scale_fill_manual(values = col) + 
  facet_grid(variable ~ substitution, scales = "free_y") + ylab("Number of mutations \n ") + 
  xlab("Motifs") +
  # coord_cartesian(ylim = c(0, ymax)) + scale_y_continuous(breaks = seq(0, ymax, 0.1)) + 
  guides(fill = FALSE) + theme_bw() + 
  theme(axis.title.y = element_text(size = 12, vjust = 1), 
        axis.text.y = element_text(size = 8), axis.title.x = element_text(size = 12), 
        axis.text.x = element_text(size = 5, angle = 90, 
                                   vjust = 0.4), strip.text.x = element_text(size = 9), 
        strip.text.y = element_text(size = 9), panel.grid.major.x = element_blank(), 
        panel.spacing.x = unit(0, "lines")) + 
   ggtitle(paste0("Mutation signatures in ", ncol(nmf_res$contribution),
                  " muscle-invasive TCGA bladder cancers")) + 
  theme(plot.title = element_text(hjust = 0.5, size = rel(1.5)))


## Figure 1B
# 按照突变数目排序
order.sample <- getSampleSummary(maf_BLCA) %>% as.data.frame() %>% .[, c("Tumor_Sample_Barcode", "total")] %>% .[order(.$total, decreasing = T), "Tumor_Sample_Barcode"] %>% as.character()
contribution.input <- function(contribution, signatures, sampleindex = NULL, signatureindex = NULL, mode){
  library(reshape2)
  if (!(mode == "relative" | mode == "absolute")) 
        stop("mode parameter should be either 'relative' or 'absolute'")
  if (is.null(sampleindex)){
    sampleindex = colnames(contribution)
  }
  if (is.null(signatureindex)){
    signatureindex = rownames(contribution)
  }
  if (mode == "relative"){ 
    contribution = contribution[signatureindex, sampleindex]
    m_contribution = melt(contribution)
    colnames(m_contribution) = c("Signature", "Sample", "Contribution")
  }else {
    if (missing(signatures))
      stop(paste("For contribution plotting in mode 'absolute':", 
                "also provide signatures matrix"))
    contribution <- contribution[signatureindex, sampleindex]
    signatures <- signatures[, signatureindex]
    total_signatures = colSums(signatures)
    abs_contribution = contribution * total_signatures
    m_contribution = melt(abs_contribution)
    colnames(m_contribution) = c("Signature", "Sample", "Contribution")
  }
  return(m_contribution)
}
relative.dat <- contribution.input(nmf_res$contribution, sampleindex = order.sample, signatureindex = c("POLE", "Unknow", "C>T_CpG", "APOBEC2", "APOBEC1"), mode = "relative")
absolute.dat <- contribution.input(contribution = nmf_res$contribution, signatures = nmf_res$signatures, sampleindex = order.sample, signatureindex = c("POLE", "Unknow", "C>T_CpG", "APOBEC2", "APOBEC1"), mode = "absolute")

p2 <- ggplot(absolute.dat, aes(x = factor(Sample), y = Contribution, 
                               fill = factor(Signature), order = Sample)) + 
  geom_bar(stat = "identity") + 
  scale_y_continuous(expand = c(0,0)) + scale_x_discrete(breaks = NULL) +
  xlab("") + ylab("Number of mutations \n ") + 
  scale_fill_manual(values = signature.col) + 
  theme_classic() +
  theme(legend.position = c(0.95, 0.7)) +
  theme(legend.background = element_blank()) + theme(legend.key = element_blank()) +
  guides(fill = guide_legend(title = NULL)) + 
  ggtitle(paste0("Signature activity across ", ncol(nmf_res$contribution),
                 " muscle-invasive TCGA bladder cancers")) + 
            theme(plot.title = element_text(hjust = 0.5, size = rel(1.5)))
        
p3 <- ggplot(relative.dat, aes(x = factor(Sample), y = Contribution, fill = factor(Signature), order = Sample)) + geom_bar(stat = "identity", position = "fill") + 
  scale_y_continuous(expand = c(0,0)) + scale_x_discrete(breaks = NULL) +
  ylab("Normalized proportion \nof signature") + xlab("") +  
  scale_fill_manual(values = signature.col) + theme_classic() +
  theme(legend.position = "none")
``` 

拼图

```{r}
p2p3 <- plot_grid(p2, p3, align = "hv", ncol = 1)
p1p2p3 <- plot_grid(p1, p2p3, labels = c("A", "B"), align = "hv", ncol = 1, rel_heights = c(3, 2))
pdf("MutationPatterns.pdf", width = 12, height = 9)
p1p2p3
dev.off()
```

![](MutationPatterns.pdf)

# Session Info

```{r}
sessionInfo()
```