---
title: "FigureYa19 Lollipop"
author: "å°ä¸«ç”»å›¾å‡ºå“"
date: "2018-7-15"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
å¾®ä¿¡ID: epigenomics  E-mail: figureya@126.com

## éœ€æ±‚æè¿°
ç”¨Rä»£ç ç”»å‡ºpaperé‡Œçš„Lollipopï¼Œæ£’æ£’ç³–ğŸ­ã€‚

åœ¨å˜‰å› å…¬ä¼—å·å›å¤â€œæ£’æ£’ç³–â€ï¼ŒæŸ¥çœ‹ä»£ç æ¥å†

![](example.png)

å‡ºè‡ª<https://www.nature.com/articles/nature22973>

## åº”ç”¨åœºæ™¯

æŸä¸ªåŸºå› çš„çªå˜å‘ç”Ÿåœ¨å“ªä¸ªé‡è¦çš„ç»“æ„åŸŸï¼Ÿæ¨æµ‹è¯¥ä½ç‚¹çš„çªå˜å½±å“äº†è›‹ç™½è´¨åŠŸèƒ½ã€‚

åœºæ™¯ä¸€ï¼šå±•ç¤ºTCGAæŸä¸ªç™Œç—‡ç±»å‹é‡ŒæŸä¸ªåŸºå› çš„çªå˜ä½ç‚¹ã€‚

åœºæ™¯äºŒï¼šå¯¹æ¯”ä¸åŒç™Œç—‡/äºšå‹çš„çªå˜ä½ç‚¹ã€‚

## ç¯å¢ƒè®¾ç½®

```r
#ä½¿ç”¨å›½å†…é•œåƒå®‰è£…åŒ…
options("repos"= c(CRAN="https://mirrors.tuna.tsinghua.edu.cn/CRAN/"))
options(BioC_mirror="http://mirrors.ustc.edu.cn/bioc/")
install.packages("BiocManager")
library(BiocManager)
install("trackViewer")
install("TCGAbiolinks")

Sys.setenv(LANGUAGE = "en") #æ˜¾ç¤ºè‹±æ–‡æŠ¥é”™ä¿¡æ¯
options(stringsAsFactors = FALSE) #ç¦æ­¢chrè½¬æˆfactor
```

```{r}
library(TCGAbiolinks)
library(maftools)
library(trackViewer)
library(RColorBrewer)
```

## ä¸‹è½½TCGAæ•°æ®

å¦‚æœä½ è¦ç”»çš„åŸºå› çªå˜ä¿¡æ¯å·²ç»ä¿å­˜åœ¨ç±»ä¼¼`easy_input.csv`çš„æ–‡ä»¶é‡Œï¼Œå°±å¯ä»¥è·³è¿‡è¿™æ­¥ï¼Œç›´æ¥è¿›å…¥â€œè¾“å…¥æ•°æ®â€ã€‚

ä¸‹è½½çªå˜ä¿¡æ¯ä½œä¸ºè¾“å…¥æ•°æ®ã€‚

```{r,message=FALSE,warning=FALSE}
mutquery <- GDCquery(project = "TCGA-LIHC",
                     data.category = "Simple Nucleotide Variation",
                     data.type = "Masked Somatic Mutation",
                     workflow.type = "MuTect2 Variant Aggregation and Masking")
GDCdownload(mutquery)
mut<-GDCprepare(mutquery)

#è¿è¡Œä¸‹é¢è¿™è¡Œï¼Œä¼šæŠŠæ‰€æœ‰åŸºå› çš„çªå˜æ•°æ®ä¿å­˜åˆ°æ–‡ä»¶é‡Œï¼Œä½ å¯ä»¥ç”¨è¿™ä¸ªæ–‡ä»¶åšæ›´å¤šçš„äº‹æƒ…ã€‚
#write.csv(mut,"TCGA_SNV.csv")

#ç”»æ£’æ£’ç³–å›¾åªéœ€è¦å…¶ä¸­çš„ä¸€ä¸ªåŸºå› ï¼Œæˆ‘ä»¬å•ç‹¬æŠ½æå‡ºæ¥ï¼Œæ­¤å¤„æå–TP53ã€‚
write.table(mut[mut$Hugo_Symbol == "TP53",],"easy_input.txt",row.names = F,quote = F,sep = "\t")
```

## è¾“å…¥æ•°æ®

éœ€è¦ç»“æ„åŸŸå’Œçªå˜ä¿¡æ¯ã€‚

### çªå˜ä¿¡æ¯

æ­¤å¤„ç›´æ¥ç”¨TCGAçš„çªå˜æ•°æ®ï¼Œæœ‰å¾ˆå¤šåˆ—ï¼Œå…¶ä¸­`RefSeq``HGVSp` `HGVSp_Short`ä¸‰åˆ—æ˜¯å¿…é¡»çš„ã€‚

å› æ­¤ï¼Œä½ ä¹Ÿå¯ä»¥åªå‡†å¤‡è¿™ä¸‰åˆ—ã€‚

```{r}
df<-read.table("easy_input.txt",as.is = T,sep = "\t",header = T)

#ç­›æ‰æ°¨åŸºé…¸æœªçªå˜çš„è¡Œ
newdf<-df[!is.na(df$HGVSp),]
```

### ç»“æ„åŸŸ

æ ¹æ®åŸºå› åï¼ˆæ­¤å¤„æ˜¯TP53ï¼‰å’Œrefseq IDä»maftoolsæä¾›çš„è›‹ç™½domainæ³¨é‡Šä¸­æå–æŸä¸ªåŸºå› çš„domainæ³¨é‡Š

```{r,message=FALSE}
gff = readRDS(file = system.file('extdata', 'protein_domains.RDs', package = 'maftools'))
refseqid <- strsplit(x = as.character(newdf$RefSeq), split = '.', fixed = TRUE)[[1]][1]
protein_inform <- gff[HGNC %in% "TP53"][refseq.ID == refseqid,]
```

### ä»çªå˜æ³¨é‡Šä¸­æå–æ°¨åŸºé…¸ä½ç½®

å…ˆå†™ä¸ªå‡½æ•°ï¼Œç”¨äºä»çªå˜æ³¨é‡Šä¸­æå–æ°¨åŸºé…¸ä½ç½®
```{r}
extractpos <- function(maf_aachange){
  prot.spl = strsplit(x = as.character(maf_aachange), split = '.', fixed = TRUE)
  prot.conv = sapply(sapply(prot.spl, function(x) x[length(x)]), '[', 1)
  pos = gsub(pattern = 'Ter.*', replacement = '',x = prot.conv)
  pos = gsub(pattern = '[[:alpha:]]', replacement = '', x = pos)
  pos = gsub(pattern = '\\*$', replacement = '', x = pos)
  pos = gsub(pattern = '^\\*', replacement = '', x = pos)
  pos = gsub(pattern = '\\*.*', replacement = '', x = pos)
  pos = as.numeric(sapply(X = strsplit(x = pos, split = '_', fixed = TRUE), FUN = function(x) x[1])) 
  aa = paste0(unlist(regmatches(maf_aachange, gregexpr("p[.].[0-9]+",maf_aachange))),"X")
  mutpos = data.frame(position = pos, mutation = maf_aachange, aa = aa, stringsAsFactors = F)
  return(mutpos[order(mutpos$pos),])
}
```

ä»çªå˜æ³¨é‡Šä¸­æå–æ°¨åŸºé…¸ä½ç½®
```{r}
pos <- extractpos(newdf$HGVSp_Short)
head(pos)
```

ç»Ÿè®¡ä¸åŒæ°¨åŸºé…¸çªå˜ç±»å‹çš„çªå˜é¢‘ç‡
```{r}
nrpos <- pos[!duplicated(pos),]
rownames(nrpos) <- nrpos$mutation
nrpos$rate <- 1
nrpos[names(table(pos$mutation)),]$rate <-table(pos$mutation)
head(nrpos)
```

## å¼€å§‹ç”»å›¾
```{r,message=FALSE}
#é¦–å…ˆå°†ä¸Šé¢è·å¾—çš„è›‹ç™½domainæ³¨é‡Šä¿¡æ¯å†™å…¥ä¸€ä¸ªGRangeså¯¹è±¡
features <- GRanges("chr1", IRanges(start=protein_inform$Start,end=protein_inform$End,names=protein_inform$Description))
features$height <- 0.07 #domain boxé«˜
features$fill<-brewer.pal(9,"Set1")[1:3] #domain boxé¢œè‰²

#å°†çªå˜ç‚¹çš„ä¿¡æ¯å†™å…¥ä¸€ä¸ªGRangeså¯¹è±¡
sample.gr <- GRanges("chr1", IRanges(nrpos$position, width=1, names=nrpos$mutation))
sample.gr$label.parameter.rot <- 90 #labelè§’åº¦
sample.gr$label <- as.character(nrpos$rate) #ç‚¹ä¸­çš„æ ‡è®°
sample.gr$label.col <- "white" #ç‚¹ä¸­æ ‡è®°çš„é¢œè‰²
sample.gr$color <-sample.gr$border <- brewer.pal(9,"Set1")[4] #ç‚¹åŠè¿çº¿é¢œè‰²
sample.gr$score<- log2(nrpos$rate) #ç‚¹æ‰€åœ¨é«˜åº¦ï¼ˆçºµåæ ‡ï¼‰

pdf("test1.pdf",width=24,height = 5)
lolliplot(sample.gr, features,xaxis = c(protein_inform$Start,protein_inform$End),yaxis = F, ylab = F,type="circle")
dev.off()
```

![](test1.pdf)

çªå˜ç‚¹å¤ªæ‹¥æŒ¤ï¼Œè§£å†³åŠæ³•æ˜¯åˆ†ä¸Šä¸‹ä¸¤å±‚ï¼Œç‚¹å¯éšæœºåˆ†ã€‚

æ³¨æ„è¿™é‡Œä¹Ÿå¯æŒ‰ç…§ä¸´åºŠä¿¡æ¯åˆ†ç±»ï¼ˆåªèƒ½åˆ†2ç±»ï¼‰ï¼Œè¿™æ ·æ›´æœ‰æ„ä¹‰ï¼Œä¸‹é¢æŠŠç‚¹éšæœºåˆ†ä¸Šä¸‹ä¸¤å±‚ï¼š

```{r}
torb <- sample(c("top", "bottom"), length(sample.gr), replace=TRUE)
sample.gr$SNPsideID <- torb
sample.gr$color[torb=="bottom"] <- sample.gr$border[torb=="bottom"] <- brewer.pal(9,"Set1")[7]

pdf("test2.pdf",width=21)
lolliplot(sample.gr, features,xaxis = c(protein_inform$Start,protein_inform$End),yaxis = F, ylab = F,type="circle")
dev.off()
```

![](test2.pdf)

ç‚¹å†…çš„æ ‡è®°å¤ªå¤šï¼Œå»æ‰1
```{r}
labs<-sample.gr$label
labs[labs=="1"]<-""
sample.gr$label <- labs
pdf("test3.pdf",width=21)
lolliplot(sample.gr, features,xaxis = c(protein_inform$Start,protein_inform$End),yaxis = F, ylab = F,type="circle")
dev.off()
```

![](test3.pdf)

æ ¹æ®çªå˜é¢‘ç‡è®¾ç½®ç‚¹å¤§å°
```{r}
sample.gr$cex <- log10(nrpos$rate) + 0.4
pdf("test4.pdf",width=21)
lolliplot(sample.gr, features,xaxis = c(protein_inform$Start,protein_inform$End),yaxis = F, ylab = F,type="circle")
dev.off()
```

![](test4.pdf)

è¿›ä¸€æ­¥æ ¹æ®çªå˜ä½ç½®æ•´åˆçªå˜,å¦‚å°†p.Y126D å’Œ p.Y126D åˆå¹¶æˆp.Y126X

```{r}
newpos <- pos[!duplicated(pos$position),]
newpos$aachange <- newpos$mutation
rownames(newpos) <-newpos$position
duppos <- names(table(pos$position)[table(pos$position)>1])
newpos[duppos,]$aachange <- newpos[duppos,]$aa

sample.gr <- GRanges("chr1", IRanges(newpos$position, width=1, names=newpos$aachange))
sample.gr$label.parameter.rot <- 45
sample.gr$label <- as.character(table(pos$position))
sample.gr$label.col <- "white"
sample.gr$color <-sample.gr$border <- brewer.pal(9,"Set1")[4]
sample.gr$score<- log2(table(pos$position))
labs<-sample.gr$label
labs[labs=="1"]<-""
sample.gr$label <- labs
torb <- sample(c("top", "bottom"), length(sample.gr), replace=TRUE)
sample.gr$SNPsideID <- torb
sample.gr$color[torb=="bottom"] <- sample.gr$border[torb=="bottom"] <- brewer.pal(9,"Set1")[7]

pdf("test5.pdf",width=21)
lolliplot(sample.gr, features,xaxis = c(protein_inform$Start,protein_inform$End),yaxis = F, ylab = F,type="circle")
dev.off()
```

![](test5.pdf)

```{r}
sessionInfo()
```