---
title: "FigureYa274MuSiCbulkProop"
author: "小丫画图出品"
date: "2022-4-22"
output: html_document
---
欢迎关注“小丫画图”公众号，回复“小白”，看小视频，实现点鼠标跑代码。

小丫微信: epigenomics  E-mail: figureya@126.com

作者：古潇，他的更多作品看这里<https://k.koudai.com/uH4zuEWC>

小丫编辑校验

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述

用MuSic反卷积推算bulk组的细胞类型。

出自<https://www.nature.com/articles/s41467-018-08023-x>

# 应用场景

MuSic这个软件大概有三个功能：

- 第一个就是直接根据单细胞数据和bulk数据预测bulk中大概含有单细胞鉴定到的哪几种细胞类型
- 第二个是先聚类，然后根据聚类信息和输入的marker list做，感觉主观性强
- 第三个是两个单细胞数据集进行相互验证，参考意义不大

流程跟着教程跑就好，官方详细教程见：<https://xuranw.github.io/MuSiC/articles/MuSiC.html#sample-analysis-1>

跟FigureYa71ssGSEA<https://k.youshop10.com/q7HAxIw2>的相似之处：核心理念都是用单细胞的数据算出来每种细胞类型的marker gene，然后看这些基因中的哪些在bulk中高表达。

同样用到单细胞和bulk的FigureYa272scBulkCCCI<https://k.youshop10.com/=fQQh=1T>，借助scRNA-seq找到的signature genes，推断bulk RNA-seq样本的细胞类群互作网络。

更多单细胞分析画图看这里<https://k.youshop10.com/McawGY6=>

> 比较难的就是构建数据集，本文档带大家解决这个问题。

分两种情况，分别带大家跑通：

- 第一种方案：**不预分组**预测bulk数据中细胞类型占比
- 第二种方案：**基于细胞类型预分组**预测bulk数据中细胞类型占比

# 环境设置

使用国内镜像安装包

```{r eval=FALSE}
options("repos"= c(CRAN="https://mirrors.tuna.tsinghua.edu.cn/CRAN/"))
options(BioC_mirror="http://mirrors.tuna.tsinghua.edu.cn/bioconductor/")
install.packages('devtools')
devtools::install_github('xuranw/MuSiC')
BiocManager::install("glmGamPoi")
```

加载包

```{r}
library(Seurat)
library(SeuratData)
library(patchwork)
library(dplyr)
library(MuSiC)
library(pheatmap)
library(magrittr)
library(Biobase)
library(glmGamPoi)

Sys.setenv(LANGUAGE = "en") #显示英文报错信息
options(stringsAsFactors = FALSE) #禁止chr转成factor
```

# 第一种方案: 不预分组预测bulk数据中细胞类型占比

## 输入文件

### 构建单细胞数据集

`ifnb` - A Seurat object with the PBMC control/IFNB-stimulated dataset。出自<https://pubmed.ncbi.nlm.nih.gov/29227470/>，已被打包到SeuratData里，我们直接安装、加载它。

为了友好对接目前主流的Seurat单细胞分析流程，使用SeuratData包钟的ifnb数据集做示例数据，示例数据ifnb有STIM和CTRL两组。

```{r}
# InstallData("ifnb") # 找网络好点的地方安装，教育网好些
LoadData("ifnb")

# 以下为Seurat标准分析流程，拿到使用SCTransform normalized的表达矩阵
ifnb.list <- SplitObject(ifnb, split.by = "stim")

ifnb.list <- lapply(X = ifnb.list, FUN = function(x) {
  x <- PercentageFeatureSet(x, pattern = "^MT-", col.name = "percent.mt")
  x <- SCTransform(x, method = "glmGamPoi", vars.to.regress = "percent.mt", verbose = FALSE)
})

features <- SelectIntegrationFeatures(object.list = ifnb.list, nfeatures = 3000)
ifnb.list <- PrepSCTIntegration(object.list = ifnb.list, anchor.features = features)

immune.anchors <- FindIntegrationAnchors(object.list = ifnb.list, normalization.method = "SCT",
                                         anchor.features = features)
immune.combined.sct <- IntegrateData(anchorset = immune.anchors, normalization.method = "SCT")

# These are now standard steps in the Seurat workflow for visualization and clustering
immune.combined.sct <- RunPCA(immune.combined.sct, npcs = 50, verbose = FALSE)

pdf("DimHeatmap.pdf",height = 20,width = 20)
DimHeatmap(immune.combined.sct, dims = 1:50, cells = 500, balanced = TRUE,ncol = 6)
dev.off()

pdf("ElbowPlot.pdf",height = 5,width = 10)
ElbowPlot(immune.combined.sct, ndims = 50)
dev.off()
```

由于SCTransform normalized的效果很好，所以使用更多的PC能揭示更多的生物学信息，这里选择30个PC

```{r}
immune.combined.sct <- RunUMAP(immune.combined.sct, dims = 1:30, verbose = FALSE)
immune.combined.sct <- FindNeighbors(immune.combined.sct, dims = 1:30, verbose = FALSE)
immune.combined.sct <- FindClusters(immune.combined.sct, resolution = 1.2, verbose = FALSE)

pdf("DimPlot.cluster.pdf",height = 5,width = 5)
DimPlot(immune.combined.sct, label = TRUE) + NoLegend()
dev.off()

pdf("DimPlot.celltype.pdf",height = 5,width = 5)
DimPlot(immune.combined.sct, group.by = "seurat_annotations", label = TRUE) + NoLegend()
dev.off()
```


```{r}
# 提取表达矩阵并构建meta信息
gene_exprs.matrix <- as.matrix(immune.combined.sct@assays$SCT@data)

metadata <- data.frame(labelDescription= c("seurat_annotations","stim","seurat_clusters"), 
                       row.names=c("seurat_annotations","stim","seurat_clusters"))
pheno.matrix <- immune.combined.sct@meta.data[,c("seurat_annotations","stim","seurat_clusters")]

# 单细胞数据集
scRNA.eset <- ExpressionSet(assayData = data.matrix(gene_exprs.matrix), 
                           phenoData =  new("AnnotatedDataFrame", data = pheno.matrix, 
                                            varMetadata = metadata))
table(scRNA.eset$seurat_annotations)
```

### 构建bulk的ExpressionSet对象

这里随机生成一个bulk数据。

对于自己的数据，需要输入normalized之后的表达矩阵，然后构建bulk的ExpressionSet对象。

```{r}
nsample <- 10
ngene <- nrow(gene_exprs.matrix)

set.seed(1) #设置随机数重点，好重复
bulk.mattrix <- matrix(runif(ngene*nsample, min = 0, max = 1000), 
                       nrow=ngene, ncol=10)
rownames(bulk.mattrix) <- rownames(gene_exprs.matrix)
colnames(bulk.mattrix) <- paste0("sample", 1:nsample)
# bulk.mattrix一定要是矩阵格式

bulk.metadata <- data.frame(labelDescription= c("sample"), 
                            row.names=c("sample"))
bulk.metadata

bulk.pheno.matrix <- data.frame(sample= colnames(bulk.mattrix), 
                                row.names=colnames(bulk.mattrix))

# bulk数据集
bulk.eset <- ExpressionSet(assayData = data.matrix(bulk.mattrix), 
                           phenoData = new("AnnotatedDataFrame", data = bulk.pheno.matrix, 
                                           varMetadata = bulk.metadata))
```

## Estimate cell type proportions

```{r}
Est.prop <- music_prop(bulk.eset = bulk.eset, # bulk数据集
                       sc.eset = scRNA.eset,  # single cell数据集
                       clusters = 'seurat_annotations', # meta文件中细胞类型的注释列名
                       samples = 'stim', # 样本注释列名
                       select.ct = c("CD14 Mono", "pDC", "CD4 Memory T", "T activated", "CD4 Naive T",
                                    "CD8 T", "Mk", "B Activated", "B", "DC", "CD16 Mono", "NK", "Eryth")) # 选择哪些细胞类型进行后续分析，默认为所有

names(Est.prop)
```

## Figure 3e - MuSiC estimated cell type proportions

```{r}
bk <- c(seq(-2,-0.1,by=0.01),seq(0,2,by=0.01))
pheatmap(Est.prop$Est.prop.weighted,
         cluster_cols = F,cluster_rows = F,
         show_colnames=T,show_rownames=T,
         #cellheight= 10,
         fontsize=16,
         #cellwidth = 10,cellheight = 10,
         treeheight_row = 0,treeheight_col_row = 10,
         #treeheight_col = 50,
         color = c(colorRampPalette(colors = c("steelblue","white"))(length(bk)/2),
                   colorRampPalette(colors = c("white","red"))(length(bk)/2)),
         border=F,border_color = "white",
         breaks=bk,scale="row",legend_breaks=seq(-2,2,1),
         filename = "Est.celltype.proportions.heatmap.pdf")
```

热图可以清楚看到随机生成的bulk数据大部分细胞为中含有单细胞鉴定的细胞类型中的pDC

## Figure 2c - Jitter plot

```{r}
# Jitter plot of estimated cell type proportions
jitter.fig <- Jitter_Est(list(data.matrix(Est.prop$Est.prop.weighted),
                             data.matrix(Est.prop$Est.prop.allgene)),
                        method.name = c('MuSiC', 'NNLS'), title = 'Jitter plot of Est Proportions')

ggsave("Est.celltype.proportions.jitterplot.pdf",plot=jitter.fig,width =10,height = 8)
```

# 第二种方案：基于细胞类型预分组预测bulk数据中细胞类型占比

## 输入文件

这里用原文提供的数据集，已提前预分组，有自定义marker list。下载地址：

- single cell dta，<https://xuranw.github.io/MuSiC/data/Mousesubeset.rds>
- bulk RNA-seq data，<https://xuranw.github.io/MuSiC/data/Mousebulkeset.rds>
- marker gene list，<https://xuranw.github.io/MuSiC/data/IEmarkers.RData>

自己的数据构建方式参考上面第一种方案。

```{r}
# Load Mouse bulk dataset
Mouse.bulk.eset <- readRDS("Mousebulkeset.rds")
Mouse.bulk.eset

# Load EMTAB single cell dataset
Mousesub.eset <- readRDS("Mousesubeset.rds")
Mousesub.eset

levels(Mousesub.eset$cellType)
```

Produce the first step information

Notice that the single cell dataset has 16 cell types, including 2 novel cell types and a transition cell type (CD-Trans). We exclude those 3 cell types in our analysis.

这里在分析中排除三种细胞类型，用剩余的细胞类型进行下游分析

```{r}
Mousesub.basis <- music_basis(Mousesub.eset, clusters = 'cellType', samples = 'sampleID', 
                             select.ct = c('Endo', 'Podo', 'PT', 'LOH', 'DCT', 'CD-PC', 'CD-IC', 'Fib',
                                           'Macro', 'Neutro','B lymph', 'T lymph', 'NK'))
```

## Figure 3a - Cluster dendrogram

Plot the dendrogram of design matrix and cross-subject mean of realtive abundance

```{r}
pdf("Est.celltype.cluster.pdf",width =10,height = 5)
par(mfrow = c(1, 2))
d <- dist(t(log(Mousesub.basis$Disgn.mtx + 1e-6)), method = "euclidean")
# Hierarchical clustering using Complete Linkage
hc1 <- hclust(d, method = "complete" )
# Plot the obtained dendrogram
plot(hc1, cex = 0.6, hang = -1, main = 'Cluster log(Design Matrix)')
d <- dist(t(log(Mousesub.basis$M.theta + 1e-8)), method = "euclidean")
# Hierarchical clustering using Complete Linkage
hc2 <- hclust(d, method = "complete")
# Plot the obtained dendrogram
plot(hc2, cex = 0.6, hang = -1, main = 'Cluster log(Mean of RA)')
dev.off()
```

The immune cells are clustered together and the kidney specific cells are clustered together. Notice that DCT and PT are within the same high-level grouping. The cut-off is user determined. Here we cut 13 cell types into 4 groups:

- Group 1: Neutro
- Group 2: Podo
- Group 3: Endo, CD-PC, CD-IC, LOH, DCT, PT
- Group 4: Fib, Macro, NK, B lymph, T lymph

The tree-guided recursive estimation for mouse kidney analysis includes 2 steps:

- Step 1. Estimate proportions of each high level cluster; (p1,p2,p3,p4)
- Step 2. Estimate cell type proportions within each cluster; (p31,p32,.,p36,p41,.,p45)

```{r}
### 这里需要根据聚类结果将细胞类型划分到不同group
# We manually specify the cluster and annotated single cell data with cluster information.
clusters.type = list(C1 = 'Neutro', C2 = 'Podo', C3 = c('Endo', 'CD-PC', 'LOH', 'CD-IC', 'DCT', 'PT'), C4 = c('Macro', 'Fib', 'B lymph', 'NK', 'T lymph'))

cl.type = as.character(Mousesub.eset$cellType)

for(cl in 1:length(clusters.type)){
  cl.type[cl.type %in% clusters.type[[cl]]] = names(clusters.type)[cl]
}
pData(Mousesub.eset)$clusterType = factor(cl.type, levels = c(names(clusters.type), 'CD-Trans', 'Novel1', 'Novel2'))

# 13 selected cell types
s.mouse = unlist(clusters.type)
s.mouse

### 自定义在C3(上皮细胞)和C4(免疫细胞)的marker gene，需根据自己的研究内容和数据自定义
(load("IEmarkers.RData"))
# This RData file provides two vectors of gene names Epith.marker and Immune.marker

# We now construct the list of group marker
IEmarkers = list(NULL, NULL, 
                 Epith.marker, Immune.marker) # marker gene
names(IEmarkers) = c('C1', 'C2', 'C3', 'C4')
# The name of group markers should be the same as the cluster names
```

## Estimate cell type proportions

```{r}
Est.mouse.bulk <- music_prop.cluster(bulk.eset = Mouse.bulk.eset, # bulk
                                     sc.eset = Mousesub.eset, # single cell
                                     group.markers = IEmarkers, clusters = 'cellType', 
                                     groups = 'clusterType', samples = 'sampleID', 
                                     clusters.type = clusters.type)
```

## Figure 3e - MuSiC estimated cell type proportions

```{r}
plot.df2 <- Est.mouse.bulk$Est.prop.weighted.cluster

library(pheatmap)
bk <- c(seq(-2,-0.1,by=0.01),seq(0,2,by=0.01))
pheatmap(plot.df2,
         cluster_cols = F,cluster_rows = F,
         show_colnames=T,show_rownames=T,
         #cellheight= 10,
         fontsize=16,
         #cellwidth = 10,cellheight = 10,
         treeheight_row = 0,treeheight_col_row = 10,
         #treeheight_col = 50,
         color = c(colorRampPalette(colors = c("steelblue","white"))(length(bk)/2),
                   colorRampPalette(colors = c("white","red"))(length(bk)/2)),
         border=F,border_color = "white",
         breaks=bk,scale="row",legend_breaks=seq(-2,2,1),
         filename = "Est.cluster.celltype.proportions.heatmap.pdf")
dev.off()
```

预测结果：该bulk数据中PT占据比例最大

# Session Info

```{r}
sessionInfo()
```